Index: rt/databinding/jaxb/src/main/java/org/apache/cxf/jaxb/JAXBContextInitializer.java
===================================================================
--- rt/databinding/jaxb/src/main/java/org/apache/cxf/jaxb/JAXBContextInitializer.java	(revision 1430123)
+++ rt/databinding/jaxb/src/main/java/org/apache/cxf/jaxb/JAXBContextInitializer.java	(working copy)
@@ -388,7 +388,8 @@
         // must not have parameters and return type must not be void
         if (method.getReturnType() == Void.class 
             || method.getParameterTypes().length != 0
-            || method.getDeclaringClass().equals(Throwable.class)
+            || (method.getDeclaringClass().equals(Throwable.class)
+            && !("getMessage".equals(method.getName())))
             || !(method.getName().startsWith("get")
                     || method.getName().startsWith("is"))) {
             return false;
